conda/content_trust/__init__.py

```diff
@@ -0,0 +1,238 @@
+            try:
+                int(basename(path).split(".")[0])
+            except ValueError:
+                # prefix is not an int and is consequently an invalid file, skip to the next
+                pass
+            else:
+                log.info(f"Loading root metadata from {path}.")
+                trusted = load_metadata_from_file(path)
+                break
+        else:
+            log.debug(
+                f"No root metadata in {context.av_data_dir}. " "Using built-in root metadata."
+            )
+
+        # Refresh trust root metadata
+        while True:
+            # TODO: caching mechanism to reduce number of refresh requests
+            fname = f"{trusted['signed']['version'] + 1}.root.json"
+            path = join(context.av_data_dir, fname)
+
+            try:
+                # TODO: support fetching root data with credentials
+                untrusted = self._fetch_channel_signing_data(
+                    context.signing_metadata_url_base,
+                    fname,
+                )
+
+                verify_root(trusted, untrusted)
+            except HTTPError as err:
+                # HTTP 404 implies no updated root.json is available, which is
+                # not really an "error" and does not need to be logged.
+                if err.response.status_code != 404:
+                    log.error(err)
+                break
+            except Exception as err:
+                # TODO: more error handling
+                log.error(err)
+                break
+            else:
+                # New trust root metadata checks out
+                trusted = untrusted
+                write_metadata_to_file(trusted, path)
+
+        return trusted
+
+    # FUTURE: Python 3.8+, replace with functools.cached_property
+    @property
+    @lru_cache(maxsize=None)
+    def key_mgr(self):
+        trusted = None
+
+        # Refresh key manager metadata
+        fname = KEY_MGR_FILE
+        path = join(context.av_data_dir, fname)
+
+        try:
+            untrusted = self._fetch_channel_signing_data(
+                context.signing_metadata_url_base,
+                KEY_MGR_FILE,
+            )
+
+            verify_delegation("key_mgr", untrusted, self.trusted_root)
+        except (ConnectionError, HTTPError) as err:
+            log.warn(err)
+        except Exception as err:
+            # TODO: more error handling
+            raise
+            log.error(err)
+        else:
+            # New key manager metadata checks out
+            trusted = untrusted
+            write_metadata_to_file(trusted, path)
+
+        # If key_mgr is unavailable from server, fall back to copy on disk
+        if not trusted and exists(path):
+            trusted = load_metadata_from_file(path)
+
+        return trusted
+
+    @staticmethod
+    def _fetch_channel_signing_data(signing_data_url, filename, etag=None, mod_stamp=None):
+        if not context.ssl_verify:
```

Comment:
This doesn't need to be a staticmethod and would be better suited to be a regular instance method to make use of standard OOP patterns.

Reply:
I opted for a `staticmethod` to clearly indicate that this is truly a helper function and doesn't depend on the current instance of `self`.

What standard OOP patterns are you referring to?

Reply:
Standard OOP patterns like subclassing, for example. If we were to decide in the future to make fetching the signing data pluggable, which would require refactoring this part of the functionality, we'd have to change the signature of this method.

A good example that I think showcases this already is the use of the `CondaSession` in the method, which could easily be decomposed into an own method to separate data fetching from data processing and making this class a little more future-proof. Your intentions regarding emphasizing, that this is a helper method would still work (that's a good pattern), by using the leading underscore in the method name.

Reply:
You're right, I didn't realize the session idea was in this method